src/lib.rs

```diff
@@ -193,6 +221,42 @@ pub fn version_meta_for(verbose_version_string: &str) -> Result<VersionMeta> {
         }
     };
 
+    let llvm_version = if let Some(&line) = out.get(idx) {
+        let llvm_version = expect_prefix(line, "LLVM version: ")?;
+        fn parse_part(part: &str) -> Result<u64> {
+            if part == "0" {
+                Ok(0)
+            } else if part.is_empty()
+                || part.as_bytes()[0] == b'0'
+                || part.find(|ch: char| !ch.is_ascii_digit()).is_some()
+            {
+                Err(Error::UnexpectedVersionFormat)
+            } else {
+                part.parse().map_err(|_| Error::UnexpectedVersionFormat)
+            }
+        }
+
+        let mut parts = llvm_version.split('.');
+        let major = parse_part(parts.next().unwrap())?;
+        let mut minor = 0;
+        if let Some(s) = parts.next() {
+            minor = parse_part(s)?;
+            if parts.next().is_some() {
+                return Err(Error::UnexpectedVersionFormat);
+            }
+            if major >= 4 && minor != 0 {
+                // only LLVM versions earlier than 4.0 can have non-zero minor versions
+                return Err(Error::UnexpectedVersionFormat);
+            }
+        } else if major < 4 {
+            // LLVM versions earlier than 4.0 have significant minor versions, so require the minor version in this case.
+            return Err(Error::UnexpectedVersionFormat);
+        }
+        Some(LLVMVersion { major, minor })
+    } else {
+        None
+    };
```

Comment:
I would propose we move this code into an `LLVMVersion::from_str() -> Result<Self, Error>` static method. Also, I'd suggest using `splitn()` rather than `split()` and ideally structuring the rest of the code as a `map()` combinator call on the result of the `splitn()` return value. I think this should obviate the need for the nested function.

Reply:
I followed your suggestion, though I don't think `splitn` would be best since I also added more specific errors. I ended up changing it back to `split` since it makes the code clearer.